Generation of picosvg font fails if a composite is encountered

```
# in https://github.com/googlefonts/color-fonts
nanoemoji --config config/noto_noflags-picosvg.toml
I0804 18:48:29.073715 139698156504896 nanoemoji.py:312] Wrote color-fonts/build/config.toml
I0804 18:48:29.074158 139698156504896 nanoemoji.py:338] Generating build.ninja
I0804 18:48:30.251909 139698156504896 nanoemoji.py:366] ninja -C color-fonts/build
ninja: Entering directory `color-fonts/build'
[3323/3323] color-fonts/venv/bin/python3 -m nanoemoji.write_font -v 0  --config config.toml @noto_noflags-picosvg.ttf.rsp
FAILED: noto_noflags-picosvg.ttf 
color-fonts/venv/bin/python3 -m nanoemoji.write_font -v 0  --config config.toml @noto_noflags-picosvg.ttf.rsp
I0804 18:49:22.683311 140083813922624 __init__.py:213] Pre-processing glyphs
I0804 18:49:22.701976 140083813922624 base.py:178] Running DecomposeComponentsFilter on noto_noflags-picosvg-Regular
I0804 18:49:22.703180 140083813922624 base.py:178] Running CubicToQuadraticFilter on noto_noflags-picosvg-Regular
I0804 18:49:22.704444 140083813922624 __init__.py:234] Building OpenType tables
I0804 18:49:23.219351 140083813922624 postProcessor.py:279] Setting post.formatType = 3.0
Traceback (most recent call last):
  File "/usr/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "nanoemoji/src/nanoemoji/write_font.py", line 653, in <module>
    app.run(main)
  File "color-fonts/venv/lib/python3.9/site-packages/absl/app.py", line 312, in run
    _run_main(main, args)
  File "color-fonts/venv/lib/python3.9/site-packages/absl/app.py", line 258, in _run_main
    sys.exit(main(argv))
  File "nanoemoji/src/nanoemoji/write_font.py", line 647, in main
    ufo, ttfont = _generate_color_font(font_config, inputs)
  File "nanoemoji/src/nanoemoji/write_font.py", line 613, in _generate_color_font
    ttfont = _make_ttfont(config, ufo, color_glyphs)
  File "nanoemoji/src/nanoemoji/write_font.py", line 216, in _make_ttfont
    _COLOR_FORMAT_GENERATORS[config.color_format].apply_ttfont(
  File "nanoemoji/src/nanoemoji/write_font.py", line 122, in <lambda>
    lambda *args: _svg_ttfont(*args, picosvg=True, compressed=False),
  File "nanoemoji/src/nanoemoji/write_font.py", line 565, in _svg_ttfont
    make_svg_table(config, ttfont, color_glyphs, picosvg, compressed)
  File "nanoemoji/src/nanoemoji/svg.py", line 547, in make_svg_table
    doc_list = _picosvg_docs(config, ttfont, color_glyphs)
  File "nanoemoji/src/nanoemoji/svg.py", line 482, in _picosvg_docs
    _add_glyph(svg, color_glyph, reuse_cache)
  File "nanoemoji/src/nanoemoji/svg.py", line 399, in _add_glyph
    raise ValueError(f"What do we do with {context}")
ValueError: What do we do with PaintTraverseContext(path=(), paint=PaintComposite(mode=<CompositeMode.SRC_IN: 5>, source=PaintColrLayers(layers=(PaintGlyph(glyph='M97.72,44.18 C99.03,44.18 100.09,45.36 100.09,46.82 C100.09,48.27 99.03,49.46 97.72,49.46 C96.41,49.46 95.35,48.28 95.35,46.82 C95.35,45.36 96.41,44.18 97.72,44.18 M97.72,42.18 C95.31,42.18 93.35,44.26 93.35,46.82 C93.35,49.38 95.31,51.46 97.72,51.46 C100.13,51.46 102.09,49.38 102.09,46.82 C102.09,44.26 100.13,42.18 97.72,42.18 Z', paint=PaintSolid(color=Color(red=66, green=66, blue=66, alpha=1.0))), PaintGlyph(glyph='M30.78,44.18 C32.09,44.18 33.15,45.36 33.15,46.82 C33.15,48.27 32.09,49.46 30.78,49.46 C29.47,49.46 28.41,48.28 28.41,46.82 C28.42,45.36 29.48,44.18 30.78,44.18 M30.78,42.18 C28.37,42.18 26.41,44.26 26.41,46.82 C26.41,49.38 28.37,51.46 30.78,51.46 C33.19,51.46 35.15,49.38 35.15,46.82 C35.15,44.26 33.2,42.18 30.78,42.18 Z', paint=PaintSolid(color=Color(red=66, green=66, blue=66, alpha=1.0))), PaintGlyph(glyph='M63.85,18.5 C65.16,18.5 66.22,19.68 66.22,21.14 C66.22,22.59 65.16,23.78 63.85,23.78 C62.54,23.78 61.48,22.6 61.48,21.14 C61.48,19.68 62.54,18.5 63.85,18.5 M63.85,16.5 C61.44,16.5 59.48,18.58 59.48,21.14 C59.48,23.7 61.44,25.78 63.85,25.78 C66.26,25.78 68.22,23.7 68.22,21.14 C68.22,18.58 66.26,16.5 63.85,16.5 Z', paint=PaintSolid(color=Color(red=66, green=66, blue=66, alpha=1.0))))), backdrop=PaintSolid(color=Color(red=0, green=0, blue=0, alpha=0.2))), transform=Affine2D(a=1, b=0, c=0, d=1, e=0, f=0))
ninja: build stopped: subcommand failed.
Traceback (most recent call last):
  File "color-fonts/venv/bin/nanoemoji", line 11, in <module>
    load_entry_point('nanoemoji', 'console_scripts', 'nanoemoji')()
  File "nanoemoji/src/nanoemoji/nanoemoji.py", line 376, in main
    app.run(_run)
  File "color-fonts/venv/lib/python3.9/site-packages/absl/app.py", line 312, in run
    _run_main(main, args)
  File "color-fonts/venv/lib/python3.9/site-packages/absl/app.py", line 258, in _run_main
    sys.exit(main(argv))
  File "nanoemoji/src/nanoemoji/nanoemoji.py", line 367, in _run
    subprocess.run(ninja_cmd, check=True)
  File "/usr/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['ninja', '-C', 'color-fonts/build']' returned non-zero exit status 1.
Traceback (most recent call last):
  File "color-fonts/build.py", line 61, in <module>
    main()
  File "color-fonts/build.py", line 50, in main
    subprocess.run(cmd, check=True)
  File "/usr/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '('nanoemoji', '--config', 'config/noto_noflags-picosvg.toml')' returned non-zero exit status 1.
```

--- a/src/nanoemoji/color_glyph.py
+++ b/src/nanoemoji/color_glyph.py
@@ -377,9 +377,10 @@ def _painted_layers(
             assert {"opacity"} == set(
                 context.element.attrib.keys()
             ), f"{debug_hint} {context.path} only attribute should be opacity. Found {context.element.attrib.keys()}"
+            # insert reversed to undo the reversed at the top of loop
             paint = PaintComposite(
                 mode=CompositeMode.SRC_IN,
-                source=PaintColrLayers(tuple(nodes)),
+                source=PaintColrLayers(tuple(reversed(nodes))),
                 backdrop=PaintSolid(Color(0, 0, 0, opacity)),
             )
             nodes = [paint]

--- a/src/nanoemoji/paint.py
+++ b/src/nanoemoji/paint.py
@@ -117,7 +117,9 @@ class Paint:
             )
             for paint in context.paint.children():
                 frontier.append(
-                    PaintTraverseContext(context.path + (self,), paint, transform)
+                    PaintTraverseContext(
+                        context.path + (context.paint,), paint, transform
+                    )
                 )
 
     def children(self) -> Iterable["Paint"]:

--- a/src/nanoemoji/svg.py
+++ b/src/nanoemoji/svg.py
@@ -18,10 +18,12 @@ import dataclasses
 from io import BytesIO
 from fontTools import ttLib
 from lxml import etree  # pytype: disable=import-error
+from nanoemoji.colors import Color
 from nanoemoji.color_glyph import ColorGlyph
 from nanoemoji.config import FontConfig
 from nanoemoji.disjoint_set import DisjointSet
 from nanoemoji.paint import (
+    CompositeMode,
     Extend,
     Paint,
     PaintSolid,
@@ -130,10 +132,15 @@ def _inter_glyph_reuse_key(
     return InterGlyphReuseKey(view_box, paint)
 
 
-def _apply_solid_paint(svg_path: etree.Element, paint: PaintSolid):
-    svg_path.attrib["fill"] = paint.color.opaque().to_string()
+def _apply_solid_paint(el: etree.Element, paint: PaintSolid):
+    if etree.QName(el.tag).localname == "g":
+        assert paint.color.opaque() == Color.fromstring(
+            "black"
+        ), "Unexpected color choice"
+    if paint.color.opaque() != Color.fromstring("black"):
+        el.attrib["fill"] = paint.color.opaque().to_string()
     if paint.color.alpha != 1.0:
-        svg_path.attrib["opacity"] = _ntos(paint.color.alpha)
+        el.attrib["opacity"] = _ntos(paint.color.alpha)
 
 
 def _apply_gradient_paint(
@@ -281,16 +288,25 @@ def _map_gradient_coordinates(paint: Paint, affine: Affine2D) -> Paint:
     raise TypeError(type(paint))
 
 
+def _is_svg_supported_composite(paint: Paint) -> bool:
+    # Only simple group opacity for now because that's all we produce in color_glyph.py
+    return (
+        isinstance(paint, PaintComposite)
+        and paint.mode == CompositeMode.SRC_IN
+        and isinstance(paint.backdrop, PaintSolid)
+    )
+
+
 def _apply_paint(
     svg_defs: etree.Element,
-    svg_path: etree.Element,
+    el: etree.Element,
     paint: Paint,
     upem_to_vbox: Affine2D,
     reuse_cache: ReuseCache,
     transform: Affine2D = Affine2D.identity(),
 ):
     if isinstance(paint, PaintSolid):
-        _apply_solid_paint(svg_path, paint)
+        _apply_solid_paint(el, paint)
     elif isinstance(paint, (PaintLinearGradient, PaintRadialGradient)):
         # Gradient paint coordinates are in UPEM space, we want them in SVG viewBox
         # so that they match the SVGPath.d coordinates (that we copy unmodified).
@@ -300,11 +316,11 @@ def _apply_paint(
             transform = Affine2D.compose_ltr(
                 (upem_to_vbox.inverse(), transform, upem_to_vbox)
             )
-        _apply_gradient_paint(svg_defs, svg_path, paint, reuse_cache, transform)
+        _apply_gradient_paint(svg_defs, el, paint, reuse_cache, transform)
     elif is_transform(paint):
         transform @= paint.gettransform()
         child = paint.paint  # pytype: disable=attribute-error
-        _apply_paint(svg_defs, svg_path, child, upem_to_vbox, reuse_cache, transform)
+        _apply_paint(svg_defs, el, child, upem_to_vbox, reuse_cache, transform)
     else:
         raise NotImplementedError(type(paint))
 
@@ -330,7 +346,7 @@ def _add_glyph(svg: SVG, color_glyph: ColorGlyph, reuse_cache: ReuseCache):
     upem_to_vbox = vbox_to_upem.inverse()
 
     # copy the shapes into our svg
-    el_by_paint = {}
+    el_by_path = {(): svg_g}
     complete_paths = set()
     for root in color_glyph.painted_layers:
         for context in root.breadth_first():
@@ -338,8 +354,12 @@ def _add_glyph(svg: SVG, color_glyph: ColorGlyph, reuse_cache: ReuseCache):
                 continue
 
             parent_el = svg_g
-            if context.path:
-                parent_el = context.path[-1]
+            path = context.path
+            while path:
+                if path in el_by_path:
+                    parent_el = el_by_path[path]
+                    break
+                path = path[:-1]
 
             if isinstance(context.paint, PaintGlyph):
                 paint_glyph = cast(PaintGlyph, context.paint)
@@ -394,7 +414,19 @@ def _add_glyph(svg: SVG, color_glyph: ColorGlyph, reuse_cache: ReuseCache):
                     parent_el.append(el)  # pytype: disable=attribute-error
                     reuse_cache.shapes[reuse_key] = el
 
+                # don't update el_by_path because we're delaring this path complete
                 complete_paths.add(context.path + (context.paint,))
+
+            elif isinstance(context.paint, PaintColrLayers):
+                pass
+
+            elif isinstance(context.paint, PaintSolid):
+                _apply_solid_paint(parent_el, context.paint)
+
+            elif _is_svg_supported_composite(context.paint):
+                el = etree.SubElement(parent_el, f"{{{svg_meta.svgns()}}}g")
+                el_by_path[context.path + (context.paint,)] = el
+
             else:
                 raise ValueError(f"What do we do with {context}")
 

--- a/src/nanoemoji/svg.py
+++ b/src/nanoemoji/svg.py
@@ -414,7 +414,7 @@ def _add_glyph(svg: SVG, color_glyph: ColorGlyph, reuse_cache: ReuseCache):
                     parent_el.append(el)  # pytype: disable=attribute-error
                     reuse_cache.shapes[reuse_key] = el
 
-                # don't update el_by_path because we're delaring this path complete
+                # don't update el_by_path because we're declaring this path complete
                 complete_paths.add(context.path + (context.paint,))
 
             elif isinstance(context.paint, PaintColrLayers):

--- a/tests/color_glyph_test.py
+++ b/tests/color_glyph_test.py
@@ -464,15 +464,15 @@ def _round_coords(paint, prec=5):
                     source=PaintColrLayers(
                         layers=(
                             PaintGlyph(
-                                glyph="M5,25 L105,25 L105,125 L5,125 Z M15,35 L15,115 L95,115 L95,35 Z",
-                                paint=PaintSolid(
-                                    color=Color(red=0, green=0, blue=255, alpha=1.0)
-                                ),
-                            ),
-                            PaintGlyph(
                                 glyph="M10,30 L100,30 L100,120 L10,120 L10,30 Z",
                                 paint=PaintSolid(
                                     color=Color(red=255, green=0, blue=0, alpha=1.0)
+                                ),
+                            ),
+                            PaintGlyph(
+                                glyph="M5,25 L105,25 L105,125 L5,125 Z M15,35 L15,115 L95,115 L95,35 Z",
+                                paint=PaintSolid(
+                                    color=Color(red=0, green=0, blue=255, alpha=1.0)
                                 ),
                             ),
                         ),

--- a/tests/group_opacity_picosvg.ttx
+++ b/tests/group_opacity_picosvg.ttx
@@ -1,0 +1,63 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ttFont sfntVersion="\x00\x01\x00\x00">
+
+  <GlyphOrder>
+    <!-- The 'id' attribute is only for humans; it is ignored when parsed. -->
+    <GlyphID id="0" name=".notdef"/>
+    <GlyphID id="1" name=".space"/>
+    <GlyphID id="2" name="e000"/>
+  </GlyphOrder>
+
+  <hmtx>
+    <mtx name=".notdef" width="0" lsb="0"/>
+    <mtx name=".space" width="100" lsb="0"/>
+    <mtx name="e000" width="100" lsb="0"/>
+  </hmtx>
+
+  <cmap>
+    <tableVersion version="0"/>
+    <cmap_format_4 platformID="0" platEncID="3" language="0">
+      <map code="0x20" name=".space"/><!-- SPACE -->
+      <map code="0xe000" name="e000"/><!-- ???? -->
+    </cmap_format_4>
+    <cmap_format_4 platformID="3" platEncID="1" language="0">
+      <map code="0x20" name=".space"/><!-- SPACE -->
+      <map code="0xe000" name="e000"/><!-- ???? -->
+    </cmap_format_4>
+  </cmap>
+
+  <loca>
+    <!-- The 'loca' table will be calculated by the compiler -->
+  </loca>
+
+  <glyf>
+
+    <!-- The xMin, yMin, xMax and yMax values
+         will be recalculated by the compiler. -->
+
+    <TTGlyph name=".notdef"/><!-- contains no outline data -->
+
+    <TTGlyph name=".space"/><!-- contains no outline data -->
+
+    <TTGlyph name="e000"/><!-- contains no outline data -->
+
+  </glyf>
+
+  <SVG>
+
+    <svgDoc endGlyphID="2" startGlyphID="2">
+      <![CDATA[<svg xmlns="http://www.w3.org/2000/svg" version="1.1">
+  <g id="glyph2" transform="matrix(0.781 0 0 0.781 0 -100)">
+    <path d="M19,11 L61,11 L61,121 L19,121 Z M29,21 L29,111 L51,111 L51,21 Z" opacity="0.8"/>
+    <g opacity="0.6">
+      <path d="M10,30 L100,30 L100,120 L10,120 L10,30 Z" fill="red"/>
+      <path d="M5,25 L105,25 L105,125 L5,125 Z M15,35 L15,115 L95,115 L95,35 Z" fill="blue"/>
+    </g>
+    <path d="M105,50 Q105,67.475 100.288,80.04 Q94.678,95 85,95 Q75.322,95 69.712,80.04 Q65,67.475 65,50 Q65,32.525 69.712,19.96 Q75.322,5 85,5 Q94.678,5 100.288,19.96 Q105,32.525 105,50 Z M95,50 Q95,34.338 90.925,23.471 Q87.748,15 85,15 Q82.252,15 79.075,23.471 Q75,34.338 75,50 Q75,65.662 79.075,76.529 Q82.252,85 85,85 Q87.748,85 90.925,76.529 Q95,65.662 95,50 Z"/>
+  </g>
+</svg>
+]]>
+    </svgDoc>
+  </SVG>
+
+</ttFont>

--- a/tests/write_font_test.py
+++ b/tests/write_font_test.py
@@ -246,6 +246,13 @@ def test_vertical_metrics(ascender, descender, linegap):
             "transformed_components_overlap.ttx",
             {"color_format": "glyf_colr_1"},
         ),
+        # Check we can produce a group with opacity from a Paint graph
+        # https://github.com/googlefonts/nanoemoji/issues/315
+        (
+            ("group_opacity.svg",),
+            "group_opacity_picosvg.ttx",
+            {"color_format": "picosvg", "pretty_print": True},
+        ),
     ],
 )
 def test_write_font_binary(svgs, expected_ttx, config_overrides):

